lib/kredis/types/proxy.rb

```diff
@@ -4,24 +4,26 @@ class Kredis::Types::Proxy
 
   attr_accessor :redis, :key
 
+  thread_mattr_accessor :pipeline
+
   def initialize(redis, key, **options)
     @redis, @key = redis, key
     options.each { |key, value| send("#{key}=", value) }
   end
 
-  def multi(&block)
-    # NOTE: to be removed when Redis 4 compatibility gets dropped
-    return redis.multi unless block
-
-    redis.multi do |pipeline|
-      block.call(Kredis::Types::Proxy.new(pipeline, key))
+  def multi(*args, **kwargs, &block)
+    redis.multi(*args, **kwargs) do |pipeline|
+      self.pipeline = pipeline
+      block.call
+    ensure
+      self.pipeline = nil
     end
   end
 
   def method_missing(method, *args, **kwargs)
     Kredis.instrument :proxy, **log_message(method, *args, **kwargs) do
       failsafe do
-        redis.public_send method, key, *args, **kwargs
+        (pipeline || redis).public_send method, key, *args, **kwargs
```

Comment:
Noting that Proxying subclasses that (mistakenly, perhaps) talk to `redis` directly will need to take care to check for `pipeline` presence as well.

Perhaps we can coalesce "current pipeline or redis" into the `redis` method itself, e.g.

```diff
diff --git a/lib/kredis/types/proxy.rb b/lib/kredis/types/proxy.rb
index 172ff2f..12dc26d 100644
--- a/lib/kredis/types/proxy.rb
+++ b/lib/kredis/types/proxy.rb
@@ -2,7 +2,9 @@ class Kredis::Types::Proxy
   require_relative "proxy/failsafe"
   include Failsafe
 
-  attr_accessor :redis, :key
+  attr_accessor :key
+
+  thread_mattr_accessor :pipeline
 
   def initialize(redis, key, **options)
     @redis, @key = redis, key
@@ -27,6 +29,10 @@ class Kredis::Types::Proxy
   end
 
   private
+    def redis
+      pipeline || @redis
+    end
+
     def log_message(method, *args, **kwargs)
       args      = args.flatten.reject(&:blank?).presence
       kwargs    = kwargs.reject { |_k, v| v.blank? }.presence
```

I'd _think_ we could remove the `:redis` attr from Proxying, too. It's a bit of a foot-gun: it can't be safely called without knowledge of whether there's an open MULTI transaction—knowledge which is not available to Proxying subclasses or their callers:
```diff
diff --git a/lib/kredis/types/proxying.rb b/lib/kredis/types/proxying.rb
index a7c0e55..3607f0f 100644
--- a/lib/kredis/types/proxying.rb
+++ b/lib/kredis/types/proxying.rb
@@ -1,14 +1,14 @@
 require "active_support/core_ext/module/delegation"
 
 class Kredis::Types::Proxying
-  attr_accessor :proxy, :redis, :key
+  attr_accessor :proxy, :key
 
   def self.proxying(*commands)
     delegate *commands, to: :proxy
   end
 
   def initialize(redis, key, **options)
-    @redis, @key = redis, key
+    @key = key
     @proxy = Kredis::Types::Proxy.new(redis, key)
     options.each { |key, value| send("#{key}=", value) }
   end
```

Reply:
Those further changes make good sense to me, thanks @jeremy.